packages/sources/s3-csv-reader/src/transport/csv.ts

```diff
@@ -0,0 +1,146 @@
+      const errorMessage = e instanceof Error ? e.message : 'Unknown error occurred'
+      logger.error(e, errorMessage)
+      response = {
+        statusCode: 502,
+        errorMessage,
+        timestamps: {
+          providerDataRequestedUnixMs: 0,
+          providerDataReceivedUnixMs: 0,
+          providerIndicatedTimeUnixMs: undefined,
+        },
+      }
+    }
+    await this.responseCache.write(this.name, [{ params: param, response }])
+  }
+
+  async _handleRequest(param: RequestParams): Promise<AdapterResponse<TransportTypes['Response']>> {
+    const { bucket, key, headerRow, resultColumn, matcherColumn, matcherValue } = param
+    const providerDataRequestedUnixMs = Date.now()
+    const csvFileAsStr = await getFileFromS3(this.s3Client, bucket, key)
+    const answer = this.findValueInCSV(
+      csvFileAsStr,
+      headerRow,
+      matcherColumn,
+      matcherValue,
+      resultColumn,
+    )
+
+    return {
+      result: answer,
+      data: {
+        result: answer,
+      },
+      statusCode: 200,
+      timestamps: {
+        providerDataRequestedUnixMs,
+        providerDataReceivedUnixMs: Date.now(),
+        providerIndicatedTimeUnixMs: undefined,
+      },
+    }
+  }
+
+  // csvFileAsStr: CSV file as string as received from S3
+  // headerRow: 1-indexed row of the CSV file to use as the header row
+  // matcherColumn: field to match with `matcherValue` to find the answer row
+  // matcherValue: value of field `matcherColumn` used to find the answer row
+  // resultColumn: header field containing the answer in matcher row
+  findValueInCSV(
+    csvFileAsStr: string,
+    headerRow: number,
+    matcherColumn: string,
+    matcherValue: string,
+    resultColumn: string,
+  ): number {
+    let parsed: Record<string, string>[]
+    try {
+      // from_line is 1-indexed
+      // columns: true sets first line as object fields rather than 2d arrays
+      parsed = parse(csvFileAsStr, {
+        columns: true,
+        from_line: headerRow,
+      })
+    } catch (error) {
+      throw new Error(`Error parsing CSV file: ${error}`)
+    }
+
+    if (parsed.length === 0) {
+      throw new Error('CSV file is empty')
+    }
+
+    const parsedHeaderRow = parsed[0]
+    // validate CSV contains headers matcherColumn and resultColumn
+    if (!(matcherColumn in parsedHeaderRow)) {
+      throw new Error(`CSV file does not contain column header ${matcherColumn}`)
+    }
+    if (!(resultColumn in parsedHeaderRow)) {
+      throw new Error(`CSV file does not contain column header ${resultColumn}`)
+    }
+
+    // find correct row using matcher
+    const matchingRows = parsed.filter((row) => row[matcherColumn] == matcherValue)
+    if (matchingRows.length === 0) {
+      throw new Error(`CSV file does not contain row where ${matcherColumn} == ${matcherValue}`)
+    }
+    if (matchingRows.length > 1) {
+      throw new Error(`CSV file contains multiple rows where ${matcherColumn} == ${matcherValue}`)
```

Comment:
yeah, the point-at-which-to-stop is a little hard to land on
At this point I'm leaning towards as simple as possible for this iteration since we know we may need to rebuild a CSV solution from scratch for distribution across NOPs.